### DISCUSSION TITLE
How to develop against a local version of glsp.client?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1204] [Q&A]
I try to test this [Pullrequest](https://github.com/eclipse-glsp/glsp-client/pull/306#issuecomment-1888833834) against my own project and I want to give some feedback.

For this I need to build the `glsp-client/packages/client` and switch to version 

```
  "dependencies": {
    "@eclipse-glsp/client": "2.1.0-next"
  }
```

I understand that `yarn link` should be my friend here. I also found [this bash script](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/configs/local-linking.sh) and adapted it to my own project. But nothing is working :-(

I will break down the problem to a very simple scenario. 
I want to build my 'model module' which only has only the dependency to `"@eclipse-glsp/client": "2.1.0-next"`

From the documentation of the [`yarn link`](https://classic.yarnpkg.com/lang/en/docs/cli/link/) command it should work the following way:

```
$ cd ~/git/glsp-client/packages/client
$ yarn install
```

next I call the link comando

```
$ yarn link
yarn link v1.22.19
success Registered "@eclipse-glsp/client".
info You can now run `yarn link "@eclipse-glsp/client"` in the projects where you want to use this package and it will be used instead.
Done in 0.04s.
```

OK - fine!

Now I switch into my model module and call the `yarn link` command again 

```
$ cd
$ cd git/open-bpmn/open-bpmn.glsp-client/open-bpmn-model/
$ yarn link "@eclipse-glsp/client"
yarn link v1.22.19
success Using linked package for "@eclipse-glsp/client".
Done in 0.03s.
```

As a result of this command the version 2.1.0-next is now part of my directory `./open-bpmn-model/@eclipse-glsp/client`

this looks all prommessing so far. But now wen call `yarn install` in my model module:

```
$ yarn install 
Couldn't find any versions for "@eclipse-glsp/client" that matches "2.1.0-next"
? Please choose a version of "@eclipse-glsp/client" from this list: (Use arrow keys)
❯ 2.1.0-next.12ca798.289 
  2.1.0-next.303 
  2.1.0-next.302 
  2.1.0-next.301 
  2.1.0-next.300 
```

Why does yarn not use my existing local version?? Even yarn install --force` has no effect :-(

**Comment by tortmayr** [comment]
Hi @rsoika,
I think the problem here is that you/we are using a mono-repository setup where all package dependency are managed in the package root.
So instead of linking the directly in your module you should do it in the root directory:
```console
cd git/open-bpmn/open-bpmn.glsp-client/
yarn link @eclipse-glsp/client`
yarn install
```
You can also check with `yarn why @eclipse-glsp/client` to see which versions of the package are resolved in your project
